Fix Vision

lzardy opened the pull request:
Fixes the issue where units would teleport to different corners of the map after re-entering vision.
Replaced NotifyEnterVision with LeaguePackets' NotifyEnterVisibilityClient. 
Replaced NotifySetHealth with LeaguePackets' NotifyEnterLocalVisibilityClient.

Fixed sometimes losing vision of a unit when a raycast stops where it started.

Fixed not being able to see through the NavGridFlag 0x40 (SEE_THROUGH, or TransparentWall), this consequently fixes not being able to attack units that are behind a see through part of the NavGrid, such as Inhibitors and Nexus'.
**NOTE: This is only applicable if the NavGrid has the flag.**
In the case were it does not, unless you decide to edit the NavGrid, use the fix suggested by #845.

When merging, close #580 and #843.

lzardy committed:
Fix Unit Teleportation on Regaining Vision

Fixes the issue where units would teleport to different corners of the map after re-entering vision. You will still see some flickering though, but instead of them only teleporting, you actually lose vision of them completely. This is due to the function IsAnythingBetween in ObjectManager.cs being poorly coded.
Replaces NotifyEnterVision with NotifyEnterVisibilityClient.
Replaces NotifySetHealth with NotifyEnterLocalVisibilityClient.

danil179 commented:
"This is due to the function IsAnythingBetween in ObjectManager.cs being poorly coded"

It doesn't seems that you changed something in this function. So this fix bypass that problem?

lzardy commented:
> "This is due to the function IsAnythingBetween in ObjectManager.cs being poorly coded"
> 
> It doesn't seems that you changed something in this function. So this fix bypass that problem?

Comment is no longer valid, issue was resolved.

lzardy committed:
Fix the rest of vision (mostly)

Fixed sometimes losing vision of a unit when the raycast stops where it started.
Fixed not being able to see through the NavGridFlag 0x40, this consequentially fixes not being able to attack units that are behind a see through part of the NavGrid, such as Inhibitors and Nexus'.
Fixed being able to see through bushes when you are not in one as well as fixed being able to see through bushes that are not in the direction of a unit getting a ray casted to it (while you are in a bush).

lzardy committed:
Fix Compare Operators

Removed IsBrush checks and added comment that bush logic should be implemented later.
Switch to correct compare operator (&&) when checking IsSeeThrough.

danil179 commented:
Try to rebase in some way that will not show changed files with all the files that got rebased. I think that if you rebase your actual commits on all the previous PRs in order (first, all the previous PRs and then your actual commits), GitHub will notice that and show only the "real" commits.

danil179 commented on PacketDefinitions420/PacketNotifier.cs:
```diff
@@ -721,23 +721,71 @@ public void NotifyLeaveVision(IGameObject o, TeamId team)
+            enterVis.LookAtPosition = new Vector3(1, 0, 0);
+            if (u is IObjAiBase)
+            {
+                _ = ((IObjAiBase)u).GetBuffs().ToList(); // buffList left for when someone decides to finish this
```
This is major style issue. Don't use `_` for naming. Why do you need the casting? I think it's redundant.

danil179 commented on GameServerLib/Packets/PacketHandlers/HandleSpawn.cs:
```diff
@@ -122,13 +122,13 @@ public override bool HandlePacket(int userId, SpawnRequest req)
             {
                 // Shop (blue team)
                  _game.PacketNotifier.NotifyStaticObjectSpawn(userId, 0xff10c6db);
```
Wrong indentation.

danil179 commented on GameServerLib/Packets/PacketHandlers/HandleSpawn.cs:
```diff
@@ -94,14 +94,14 @@ public override bool HandlePacket(int userId, SpawnRequest req)
                 else if (kv.Value is IInhibitor || kv.Value is INexus)
                 {
                     var inhibtor = (IAttackableUnit)kv.Value;
                      _game.PacketNotifier.NotifyStaticObjectSpawn(userId, inhibtor.NetId);
```
Wrong indentation.

danil179 commented on GameServerLib/Content/NavGrid.cs:
```diff
@@ -490,34 +510,33 @@ public float CastRaySqr(Vector2 origin, Vector2 destination, bool inverseRay = f
             {
-                if (IsWalkable(x1, y1) == inverseRay)
+                //TODO: Implement bush logic (preferably near here)
+                if (IsWalkable(x1, y1) == checkWalkable && IsSeeThrough(x1, y1) == checkWalkable)
```
Add comment: `// Only works for navgrids with SeeThrough flags`

danil179 commented on GameServerLib/Maps/SummonersRift.cs:
```diff
@@ -273,15 +273,14 @@ public void Init()
-            _game.ObjectManager.AddObject(new Inhibitor(_game, "ChaosInhibitor", TeamId.TEAM_PURPLE, collisionRadius, 11240, 11490, sightRange, 0xffff8f1f)); //mid
-            _game.ObjectManager.AddObject(new Inhibitor(_game, "ChaosInhibitor", TeamId.TEAM_PURPLE, collisionRadius, 13200, 11200, sightRange, 0xff26ac0f)); //bot
-
+            _game.ObjectManager.AddObject(new Inhibitor(_game, "OrderInhibitor", TeamId.TEAM_BLUE, collisionRadius, 853.368f, 3345.1645f, sightRange, 0xffd23c3e)); //top exact
```
Add comment: `// TODO: Generate exact positions from content files`

danil179 commented on GameServerCore/Packets/Interfaces/IPacketNotifier.cs:
```diff
@@ -35,8 +35,9 @@ public interface IPacketNotifier
         void NotifyEmotions(Emotions type, uint netId);
-        void NotifyEnterVision(IGameObject o, TeamId team);
-        void NotifyEnterVision(int userId, IChampion champion);
+        void NotifyEnterLocalVisibilityClient(IAttackableUnit unit, int userId = 0);
```
I don't really like that `userId` is the last argument. I get why you did that (you need default value), but still it looks weird.

lzardy replied:
Can I ignore this? I need the userId = 0 for the check if userId == 0, then broadcast. Perhaps we can discuss if EnterLocalVisibilityClient should Broadcast?

danil179 replied:
Add flag `IsBroadcast`.

danil179 commented on GameServerLib/Packets/PacketHandlers/HandleSpawn.cs:
```diff
@@ -122,13 +122,13 @@ public override bool HandlePacket(int userId, SpawnRequest req)
                 // Shop (blue team)
                  _game.PacketNotifier.NotifyStaticObjectSpawn(userId, 0xff10c6db);
-                 _game.PacketNotifier.NotifySetHealth(userId, 0xff10c6db);
+                _game.PacketNotifier.NotifyEnterLocalVisibilityClient(userId, 0xff10c6db);
```
`0xff10c6db` is some hardcoded `NetId`. Add comment: `// TODO: Generate shop NetId to avoid hard-coding`

danil179 commented on PacketDefinitions420/PacketNotifier.cs:
```diff
@@ -721,23 +721,71 @@ public void NotifyLeaveVision(IGameObject o, TeamId team)
 
-        public void NotifyEnterVision(IGameObject o, TeamId team)
+        // TODO: Fix IsAnythingBetween randomly losing vision often, which ends up spamming this as well
+        public void NotifyEnterVisibilityClient(IAttackableUnit u, TeamId team, int userId = 0)
```
Related implementation: #832

lzardy replied:
Not sure if you want me to change something, the implementation linked has many features that my PR lacks, such as SyncId and MovementData, thus my implementation can't change much.

danil179 replied:
I just wanted to relate this in GitHub (so I'll have a link).

danil179 reviewed (changes requested):
Style issues and some discussion.

lzardy committed:
Resolve PR comments

danil179 reviewed (approved):
Approved. Consider changing the `userId` to broadcast flag where needed. (if it's possible)
